README.md
<!-- badges: start -->
[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![Codecov test coverage](https://codecov.io/gh/ropensci/tidyqpcr/branch/main/graph/badge.svg)](https://codecov.io/gh/ropensci/tidyqpcr/branch/main)
[![fair-software.eu](https://img.shields.io/badge/fair--software.eu-%E2%97%8F%20%20%E2%97%8F%20%20%E2%97%8F%20%20%E2%97%8F%20%20%E2%97%8F-green)](https://fair-software.eu)
[![CII Best Practices](https://bestpractices.coreinfrastructure.org/projects/5287/badge)](https://bestpractices.coreinfrastructure.org/projects/5287)
[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/tidyqpcr)](https://cran.r-project.org/package=tidyqpcr)
[![R-CMD-check](https://github.com/ropensci/tidyqpcr/workflows/R-CMD-check/badge.svg)](https://github.com/ropensci/tidyqpcr/actions)
<!-- badges: end -->

# tidyqpcr - Quantitative PCR analysis in the tidyverse.

*Empowering scientists to conduct reproducible, flexible, and MIQE best-practice compliant quantitative PCR analysis.*

# Contents

* [Motivation](#Motivation)
	* [Empowering](#Empowering)
	* [Reproducible](#Reproducible)
	* [Flexible](#Flexible)
	* [Best-practice compliant](#Best-practice-compliant)
* [Getting started](#Getting-started)
	* [Installing tidyqpcr](#Installing-tidyqpcr)
	* [Using tidyqpcr](#Using-tidyqpcr)
* [Status](#Status)
    * [News - see NEWS.md](NEWS.md)
* [Features](#Features)
	* [Current features](#Current-features)
	* [Future priorities](#Future-priorities)
* [Contribute](#Contribute)


# Motivation

Quantitative Polymerase Chain Reaction (qPCR) is a highly adaptable experimental technique used across biology and medicine to measure the amounts of nucleic acids (DNA or RNA).
tidyqpcr is a software package for qPCR data analysis that builds on the tidyverse collection of data science tools in the R programming language.

### Empowering

tidyqpcr combines a free, open-source qPCR analysis R package with online teaching materials. 

We want our users to be able to know and understand what happens at every step of their analysis.
Users are able to know what occurs at each step as all tidyqpcr tools are open source and follow the FAIR principles - [Findable, Accessible, Interoperable, and Reusable](https://force11.org/info/the-fair-data-principles/). 
Users should also find each step understandable as we aim to produce educational resources as extensions of [data carpentry](https://datacarpentry.org/) workshops, such as [Data Analysis and Visualization in R for Ecologists](https://datacarpentry.org/R-ecology-lesson/), accessible to beginner programmers. 

### Reproducible

tidyqpcr scripts produce paper-ready figures straight from raw data with identical results across computers.

We want to promote reproducible research so collaborators, reviewers or students can easily confirm and extend results and conclusions. tidyqpcr analysis will repeat exactly on different computers, enabling scientists to share raw data and analysis scripts rather than just  processed figures. An R or R markdown script using tidyqpcr to analyse a set of qPCR data could be directly uploaded to a repository such as [figshare](https://figshare.com/), as encouraged by many journal publishers.

### Flexible

tidyqpcr follows the 'tidy' data paradigm to ensure scalability and adaptability.

We want to create a tool that is flexible enough to analyse high or low throughput experimental data whilst integrating easily into other data analyses. tidyqpcr uses powerful generic data science tools from [the tidyverse](https://www.tidyverse.org/) R package, lightly overlaid with qPCR-specific scripts. As far as possible, every object in tidyqpcr is stored as a generic tibble / data frame. Manipulating and plotting qPCR data without creating bespoke data structures allows tidyqpcr scripts to be easily integrated and scaled according to the needs of your experiments.

### Best-practice compliant

tidyqpcr encourages standardised, reliable experimental design by following the Minimum Information for Publication of Quantitative Real-Time PCR Experiments (MIQE) best practices.

We want to make it easier for scientists to produce reliable and interpretable results. The MIQE best practices are a framework to facilitate the full disclosure of all reagents, sequences, and analysis methods necessary to enable other investigators to reproduce results. The final version of tidyqpcr will, by default, request the relevant experimental conditions and assay characteristics, as described in the [MIQE guidelines](https://academic.oup.com/clinchem/article/55/4/611/5631762), to allow reviewers/readers to rigorously assess the validity of a result. See "Future Priorities" below to get updates on tidyqpcr's MIQE compliant features.


# Getting started

## Installing tidyqpcr

First install [R](https://www.r-project.org/). 

#### For Windows users

Next, you need a working installation of [Rtools](https://cran.r-project.org/bin/windows/Rtools/rtools40.html).

Jeffrey Leek made [slides on installation and testing of Rtools](http://jtleek.com/modules/01_DataScientistToolbox/02_10_rtools/).

### Install via devtools (for all R users)

Install the devtools R package, see [devtools installation instructions](https://www.r-project.org/nosvn/pandoc/devtools.html). 

```
library(devtools)
devtools::install_github("ropensci/tidyqpcr", build_vignettes = TRUE)

## Alternatively, install without building the vignettes 
## (Not recommended as vignettes contain the tutorials on using tidyqpcr)
devtools::install_github("ropensci/tidyqpcr")
```

**Note**
older versions of the remotes package automatically convert warnings to errors during installation.
Please update your remotes package to >2.3.0 in order to remove this default.

Then load tidyqpcr as a standard package:

```
library(tidyqpcr)
```

**Note**
tidyqpcr automatically imports and loads several external packages for basic functionality, including; tidy, dplyr and ggplot2.
This allows tidyqpcr to be used immediately but may cause NAMESPACE clashes if the user already has many other package libraries loaded.
Restarting the R session and loading tidyqpcr separately may solve such issues.

## Using tidyqpcr

The best place to start is by viewing the articles on the [tidyqpcr website](https://docs.ropensci.org/tidyqpcr/index.html).
Here you will find the vignettes, which offer tutorials and example data analyses including figures.
Currently there are 4 vignettes:

* [IntroDesignPlatesetup](https://docs.ropensci.org/tidyqpcr/articles/platesetup_vignette.html) - Introduction to designing an experiment and setting up a plate plan in tidyqpcr.
* [DeltaCq96wellExample](https://docs.ropensci.org/tidyqpcr/articles/deltacq_96well_vignette.html) - Example analysis of 96-well RT-qPCR data including relative quantification with delta Cq, from a real experiment.
* [MultifactorialExample](https://docs.ropensci.org/tidyqpcr/articles/multifactor_vignette.html) - Example design and analysis of a (real) multifactorial RT-qPCR experiment.
* [PrimerCalibration](https://docs.ropensci.org/tidyqpcr/articles/calibration_vignette.html) - Example design and analysis of calibrating qPCR primer sets from a (real) experimental test

To find these from your R session, enter `browseVignettes(package="tidyqpcr")`. 

Individual R functions are also documented, use R's standard help system after loading the package, e.g. `?create_blank_plate`. To see a list of all the functions and links to their help pages use `help(package="tidyqpcr")`.

A basic use case for designing a 12 well plate is given below, see [IntroDesignPlatesetup](https://docs.ropensci.org/tidyqpcr/articles/platesetup_vignette.html) for more details.

```
rowkey4 <- tibble(
  well_row = LETTERS[1:4],
  target_id = c("ACT1", "BFG2", "CDC19", "DED1")
)


colkey3 <- tibble(
  well_col = 1:3,
  sample_id = c("rep1", "rep2", "rep3"),
  prep_type = "+RT"
)

create_blank_plate(well_row = LETTERS[1:4], well_col = 1:3)

plate_plan12 <- label_plate_rowcol(
  plate = create_blank_plate(well_row = LETTERS[1:4], well_col = 1:3),
  rowkey = rowkey4,
  colkey = colkey3
)

display_plate_qpcr(plate_plan12)
```

# Status

As of June 2022, this software is fully useable, and under active development.
It is particularly good at designing qPCR experiments in microwell plates (96-well and 384-well), and at relative quantification by the delta Cq method.

[Edward Wallace](https://github.com/ewallace) wrote basic functions and documentation needed to do qPCR analysis in [the Wallace lab](https://ewallace.github.io/), then started building them into an R package.
[Sam Haynes](https://github.com/dimmestp) is actively developing, initially as part of the [eLife Open Innovation Leaders programme 2020](https://elifesciences.org/labs/fdcb6588/innovation-leaders-2020-introducing-the-cohort).

If there is a feature that you need for your work, please ask us! 

## News - see [NEWS.md](NEWS.md).

# Features 

tidyqpcr can be used to analyse qPCR data from any nucleic acid source - DNA for qPCR or ChIP-qPCR, RNA for RT-qPCR.

Currently tidyqpcr has functions that support relative quantification by the delta Cq method, but not yet absolute quantification.


## Current features

* every object is a tibble / data frame, no special data classes to learn
* lay out and display 96/384-well plates for easy experimental setup (`label_plate_rowcol`, `create_blank_plate`, ...).
* consistently describe samples and target amplicons with reserved variable names (`sample_id`, `target_id`).
* flexibly assign metadata to samples for visualisation with [ggplot2](https://ggplot2.tidyverse.org/) (see vignettes).
* read in quantification cycle (Cq) and raw data from Roche LightCycler machines with single-channel fluorescence (`read_lightcycler_1colour_cq`, `read_lightcycler_1colour_raw`).
* calibration of primer sets including estimating efficiencies and visualization of curves (`calculate_efficiency`, and see vignettes)
* visualization of amplification and melt curves (`calculate_drdt_plate`, and see vignettes)
* delta Cq: normalization/ relative quantification of Cq data to one or more reference targets by delta count method (`calculate_normcq`, `calculate_deltacq_bysampleid`)
* delta delta Cq: normalization of delta Cq data across multiple samples (`calculate_deltadeltacq_bytargetid`)

## Future priorities

* including primer efficiencies in quantification
* an open-source and tested Cq calculation function, from amplification curves
* multi-colour (hydrolysis probe) detection
* extend to 1536-well plates 
* metadata handling compatible with RDML format
* files for automatic plate loading with [Opentrons](https://opentrons.com/) and [Labcyte Echo](https://www.labcyte.com/products/liquid-handling/echo-liquid-handlers) liquid handlers.

# Comparison of qPCR R packages with respect to the MIQE guidelines

Table of package features corresponding to the essential information on qPCR validation and data analysis, that are outlined by the MIQE guidelines for publication of qPCR results. 

| MIQE Guidelines | tidyqpcr | HTqPCR | NormqPCR | qpcR | pcr |
--- | --- | --- | --- | --- | ---
Version | 0.5.0 | 1.48.0 | 1.40.0 | 1.4.1  | 1.2.2
For SYBR Green I, Cq of the NTC | Yes + Docs | Yes + Docs | Yes + Docs | Yes | Yes
Calibration curves with slope and y intercept | Slope | No | No | Yes + Docs | Yes + Docs
PCR efficiency calculated from slope | Yes + Doc | No | Yes + Doc | Yes + Doc | Yes
r2 of calibration curve | Yes + Doc | No | No | Yes | Yes
Linear dynamic range ‡ | No | No | No | No | No
Cq variation at LOD ‡ | No | No | No | No | No
Evidence for LOD ‡ | No | No | No | No | No
If multiplex, efficiency and LOD of each assay ‡ | No | No | No | No | No

NEWS.md
## News

* June 2022, removed plot helper functions `scale_..._nice` and `scale_loglog` from tidyqpcr, because those capabilities are now available in the [scales package](https://scales.r-lib.org) using `label_log` and similar functions. Older code may need to change `scale_y_log10nice` to `scale_y_log10(labels = scales::label_log())`, for example.
* May 2022, Improvements in documentation and testing. Reorganized `display_plate` function to be more flexible, so older code will need to use `display_plate_qpcr` to ensure that `sample_id` and `target_id` info displays. Updated to v0.5.
* January 2022, Improvements in documentation and argument-checking for v0.4.
* October 2021, Unit tests now cover over 75% of tidyqpcr code.
* June 2021, [tidyqpcr blogpost in eLife labs](https://elifesciences.org/labs/f23e268f/tidyqpcr-quantitative-pcr-analysis-in-the-tidyverse)
* August 2020, relative quantification (delta delta Cq) added with function `calculate_deltadeltacq_bytargetid`, and a vignette illustrating this with a small data set from a 96-well plate.
* June 2020, upgrades that break previous code. All function and variable names have been changed to snake case, i.e. lower case with underscore. Commits up to #ee6d192 change variable and function names. tidyqpcr now uses `sample_id` for nucleic acid sample (replaces Sample or SampleID), `target_id` for primer set/ probe (replaces TargetID or Probe), `prep_type` for nucleic acid preparation type (replaces Type), and `cq` for quantification cycle (replaces Cq or Ct). It should be possible to upgrade old analysis code by (case-sensitive) search and replace. 


R/plate_functions.R
#' Create a blank plate template as a tibble (with helper functions for common plate sizes)
#'
#' For more help, examples and explanations, see the plate setup vignette:
#' \code{vignette("platesetup_vignette", package = "tidyqpcr")}
#'
#' @param well_row Vector of Row labels, usually LETTERS
#' @param well_col Vector of Column labels, usually numbers
#' @return tibble (data frame) with columns well_row, well_col, well. This
#'   contains all pairwise combinations of well_row and well_col, as well as
#'   individual well names. Both well_row and well_col are coerced to factors
#'   (even if well_col is supplied as numbers), to ensure order is consistent.
#'
#'   However, well is a character vector as that is the default behaviour of
#'   "unite", and display order doesn't matter.
#'
#'   Default value describes a full 384-well plate.
#'
#' @examples
#' create_blank_plate(well_row=LETTERS[1:2],well_col=1:3)
#' 
#' create_blank_plate_96well()
#' 
#' create_blank_plate_1536well()
#' 
#' # create blank 96-well plate with empty edge wells
#' 
#' create_blank_plate(well_row=LETTERS[2:7], well_col=2:11)
#' 
#' # create blank 1536-well plate with empty edge wells
#' 
#' full_plate_row_names <- make_row_names_lc1536()
#' 
#' create_blank_plate(well_row=full_plate_row_names[2:31], well_col=2:47)
#'
#' @family plate creation functions
#'
#' @export
#' @importFrom tibble tibble as_tibble
#' @importFrom tidyr %>%
#' @importFrom forcats as_factor
#' @importFrom rlang .data
#'
create_blank_plate <- function(well_row = LETTERS[1:16], well_col = 1:24) {
    tidyr::crossing(well_row = as_factor(well_row),
                    well_col = as_factor(well_col)) %>%
        as_tibble() %>%
        tidyr::unite("well", .data$well_row, .data$well_col, 
                     sep = "", remove = FALSE)
}

#' @describeIn create_blank_plate create blank 96-well plate
#' @export
#'
create_blank_plate_96well <- function() {
    create_blank_plate(well_row = LETTERS[1:8], well_col = 1:12)
}

#' Generates row names for the Roche Lightcycler (tm) 1536-well plates
#' 
#' Creates a vector containing 36 row names according to the labelling system 
#' used by the Roche Lightcycler (tm)
#' 
#' @return Vector of row names: Aa,Ab,Ac,Ad,Ba,...,Hd.
#' 
#' @family plate creation functions
#' 
#' @examples
#' make_row_names_lc1536()
#' 
#' @export
#'
make_row_names_lc1536 <- function() {
    paste0(rep(LETTERS[1:8], each = 4), letters[1:4])
}

#' Generates row names for the Labcyte Echo 1536-well plates 
#' 
#' Creates a vector containing 36 row names according to the 
#' labelling system used by the Labcyte Echo
#' 
#' @return Vector of row names: A,B,...,Z,AA,AB,...,AF.
#' 
#' @family plate creation functions
#' 
#' @examples
#' make_row_names_echo1536()
#' 
#' @export
#'
make_row_names_echo1536 <- function() {
    c(LETTERS[1:26], paste0("A", LETTERS[1:6]))
}

#' @describeIn create_blank_plate create blank 1536-well plate
#' @export
#'
create_blank_plate_1536well <- function(
    well_row = make_row_names_lc1536(),
    well_col = 1:48) {
    create_blank_plate(well_row, well_col)
}

#' Create a 6-value, 24-column key for plates
#'
#' Create a 24-column key with 6 values repeated over 24 plate columns.
#' Each of the 6 values is repeated over 3x +RT Techreps and 1x -RT.
#' 
#' This helps to create plate layouts with standard designs.
#'
#' @param ... Vectors of length 6 describing well contents,
#' e.g. sample_id or target_id
#' @return tibble (data frame) with 24 rows, and columns
#' well_col, prep_type, tech_rep, and supplied values.
#'
#' @examples
#' create_colkey_6_in_24(sample_id=LETTERS[1:6])
#' @family plate creation functions
#'
#' @export
#' @importFrom tibble tibble as_tibble

#' @importFrom tidyr %>%
#'
create_colkey_6_in_24 <- function(...) {
    colkey <- tibble(well_col  = factor(1:24),
                     prep_type = as_factor(c(rep("+RT", 18), rep("-RT", 6))),
                     tech_rep  = as_factor(rep(c(1, 2, 3, 1), each = 6))
                     )
    if (!missing(...)) {
        pieces6 <- list(...) %>% as_tibble()
        assertthat::assert_that(nrow(pieces6) == 6, 
                                msg = "Some input data is not of length 6")
        pieces24 <- dplyr::bind_rows(pieces6, pieces6, pieces6, pieces6)
        colkey <- dplyr::bind_cols(colkey, pieces24)
    }
    return(colkey)
}

#' Create a 4-dilution column key for primer calibration
#'
#' Creates a 24-column key for primer calibration, with 2x biol_reps and 2x
#' tech_reps, and 5-fold dilution until 5^4 of +RT; then -RT (no reverse
#' transcriptase), NT (no template) negative controls. That is a total of 6
#' versions of each sample replicate.
#'
#' @param dilution Numeric vector of length 6 describing sample dilutions
#' @param dilution_nice Character vector of length 6 with nice labels for sample
#'   dilutions
#' @param prep_type Character vector of length 6 describing type of sample (+RT,
#'   -RT, NT)
#' @param biol_rep Character vector of length 6 describing biological replicates
#' @param tech_rep Character vector of length 6 describing technical replicates
#' @return tibble (data frame) with 24 rows, and columns well_col, dilution,
#'   dilution_nice, prep_type, biol_rep, tech_rep.
#' @examples
#' create_colkey_4diln_2ctrl_in_24()
#' @family plate creation functions
#'
#' @export
#' @importFrom tibble tibble
#' @importFrom forcats as_factor
#'
create_colkey_4diln_2ctrl_in_24 <- function(
                     dilution      = c(5 ^ (0:-3), 1, 1),
                     dilution_nice = c("1x", "5x", "25x", "125x", "-RT", "NT"),
                     prep_type     = c(rep("+RT", 4), "-RT", "NT"),
                     biol_rep      = rep(c("A", "B"), each = 12,
                                        length.out = 24),
                     tech_rep      = rep(1:2, each = 6,
                                        length.out = 24)
                     ) {
    tibble(well_col = factor(1:24),
           dilution = rep(dilution, 4),
           dilution_nice = rep(dilution_nice, 4),
           prep_type = as_factor(rep(prep_type, 4)),
           biol_rep = as_factor(biol_rep),
           tech_rep = as_factor(tech_rep)
    )
}

#' Create a 6-dilution column key for primer calibration
#'
#' Creates a 24-column key for primer calibration, with 1x biol_reps and 3x
#' tech_reps, and 5-fold dilution until 5^6 of +RT; then -RT (no reverse
#' transcriptase), NT (no template) negative controls. That is a total of 8
#' versions of each replicate.
#'
#' @param dilution Numeric vector of length 8 describing sample dilutions
#' @param dilution_nice Character vector of length 8 with nice labels for sample
#'   dilutions
#' @param prep_type Character vector of length 8 describing type of sample (+RT,
#'   -RT, NT)
#' @param tech_rep Character vector of length 8 describing technical replicates
#' @return tibble (data frame) with 24 rows, and variables well_col, dilution,
#'   dilution_nice, prep_type, biol_rep, tech_rep.
#' @examples
#' create_colkey_6diln_2ctrl_in_24()
#' @family plate creation functions
#'
#' @export
#' @importFrom tibble tibble
#' @importFrom forcats as_factor
#'
create_colkey_6diln_2ctrl_in_24 <- function(
                     dilution = c(5 ^ (0:-5), 1, 1),
                     dilution_nice = c("1x", "5x", "25x", "125x",
                                    "625x", "3125x", "-RT", "NT"),
                     prep_type=c(rep("+RT", 6), "-RT", "NT"),
                     tech_rep = rep(1:3, each = 8, length.out = 24)
                     ) {
    tibble(well_col = factor(1:24),
           dilution = rep(dilution, 3),
           dilution_nice = rep(dilution_nice, 3),
           prep_type = as_factor(rep(prep_type, 3)),
           tech_rep = as_factor(tech_rep))
}

#' Create a 4-value, 16-row key for plates
#'
#' Create a 16-row key with 4 values repeated over 16 plate rows. Each of the 4
#' values is repeated over 3x +RT Techreps and 1x -RT.
#' 
#' This helps to create plate layouts with standard designs.
#'
#' @param ... Vectors of length 4 describing well contents, e.g. sample_id or
#'   target_id
#' @return tibble (data frame) with 16 rows, and variables well_row, prep_type,
#'   tech_rep, and supplied values.
#' @examples
#' create_rowkey_4_in_16(sample_id=c("sheep","goat","cow","chicken"))
#' @family plate creation functions
#'
#' @export
#' @importFrom tibble tibble as_tibble
#' @importFrom forcats as_factor
#' @importFrom tidyr %>%
#'
create_rowkey_4_in_16 <- function(...) {
    rowkey <- tibble(well_row = factor(LETTERS[1:16]),
                     prep_type = as_factor(c(rep("+RT", 12), rep("-RT", 4))),
                     tech_rep = as_factor(rep(c(1, 2, 3, 1), each = 4))
    )
    if (!missing(...)) {
        pieces4 <- list(...) %>% as_tibble()
        assertthat::assert_that(nrow(pieces4) == 4, 
                                msg = "Some input data is not of length 4")
        pieces16 <- dplyr::bind_rows(pieces4, pieces4, pieces4, pieces4)
        rowkey <- dplyr::bind_cols(rowkey, pieces16)
    }
    return(rowkey)
}

#' Create a plain 8-value, 16-row key for plates
#'
#' Create a 16-row key with 8 values repeated over 16 plate rows. No other
#' information is included by default, hence "plain".
#' 
#' This helps to create plate layouts with standard designs.
#'
#' @param ... Vectors of length 8 describing well contents, e.g. sample or
#'   probe.
#' @return tibble (data frame) with 16 rows, and variables well_col, and
#'   supplied values.
#' @examples
#' create_rowkey_8_in_16_plain(sample_id=c("me","you","them","him",
#'                                    "her","dog","cat","monkey"))
#' @family plate creation functions
#'
#' @export
#' @importFrom tibble tibble
#' @importFrom tidyr %>%
#'
create_rowkey_8_in_16_plain <- function(...) {
    rowkey <- tibble(well_row = factor(LETTERS[1:16]))
    if (!missing(...)) {
        pieces8 <- list(...) %>% as_tibble()
        assertthat::assert_that(nrow(pieces8) == 8, 
                                msg = "Some input data is not of length 8")
        pieces16 <- dplyr::bind_rows(pieces8, pieces8)
        rowkey <- dplyr::bind_cols(rowkey, pieces16)
    }
    return(rowkey)
}

#' Label a plate with sample and probe information
#'
#' For more help, examples and explanations, see the plate setup vignette:
#' \code{vignette("platesetup_vignette", package = "tidyqpcr")}
#' 
#' For worked examples of tidyqpcr analysis with 384-well plates, see:
#' \code{vignette("calibration_vignette", package = "tidyqpcr")}
#'
#' @param plate tibble (data frame) with variables well_row, well_col, well.
#'   This would usually be produced by create_blank_plate(). It is possible to
#'   include other information in additional variables.
#' @param rowkey tibble (data frame) describing plate rows, with variables
#'   well_row and others.
#' @param colkey tibble (data frame) describing plate columns, with variables
#'   well_col and others.
#' @param coercefactors if TRUE, coerce well_row in rowkey and well_col in
#'   colkey to factors
#'
#' @return tibble (data frame) with variables well_row, well_col, well, and
#'   others.
#'
#'   This tibble contains all combinations of well_row and well_col found in the
#'   input plate, and all information supplied in rowkey and colkey distributed
#'   across every well of the plate. Return plate is ordered by row well_row
#'   then column well_col.
#'
#'   Note this ordering may cause a problem if well_col is supplied as a
#'   character (1,10,11,...), instead of a factor or integer (1,2,3,...). For
#'   this reason, the function by default converts well_row in `rowkey`, and
#'   well_col in `colkey`, to factors, taking factor levels from `plate`, and
#'   messages the user.
#'   
#'   If `plate$well_col` or `plate$well_row` are not factors and coercefactors = TRUE 
#'   label_plate_rowcol will automatically convert them to factors, but will output a 
#'   warning telling users this may lead to unexpected behaviour. 
#'
#'   Other tidyqpcr functions require plate plans to contain variables
#'   sample_id, target_id, and prep_type, so `label_plate_rowcol` will message
#'   if any of these are missing. This is a message, not an error, because these
#'   variables can be added by users later.
#'
#' @examples
#' label_plate_rowcol(plate = create_blank_plate()) # returns blank plate
#' 
#' # label blank 96-well plate with empty edge wells
#' 
#' label_plate_rowcol(plate = create_blank_plate(well_row = LETTERS[2:7], 
#'                                               well_col = 2:11))
#' 
#' # label 96-well plate with sample id in rows
#' 
#' label_plate_rowcol(plate = create_blank_plate(well_row = LETTERS[1:8],
#'                                               well_col = 1:12),
#'                    rowkey = tibble(well_row = LETTERS[1:8],
#'                                    sample_id = paste0("S_",1:8)))
#' 
#' # label fraction of 96-well plate with target id in columns
#' 
#' label_plate_rowcol(plate = create_blank_plate(well_row = LETTERS[1:8],
#'                                               well_col = 1:4),
#'                    colkey = tibble(well_col = 1:4,
#'                                    target_id = paste0("T_",1:4)))
#' 
#' @family plate creation functions
#'
#' @export
#'
#' @importFrom rlang .data
#' @importFrom forcats as_factor
#'
label_plate_rowcol <- function(plate,
                               rowkey = NULL,
                               colkey = NULL,
                               coercefactors = TRUE) {
    assertthat::assert_that(
        assertthat::has_name(plate, 
                             c("well_row","well_col")))
    
    if (!is.factor(plate$well_col) & coercefactors){
        warning("plate$well_col is not a factor. Automatically generating plate$well_col factor levels. May lead to incorrect plate plans.")
        plate <- plate %>%
            dplyr::mutate(well_col = as_factor(.data$well_col))
    }
    
    if (!is.factor(plate$well_row) & coercefactors){
        warning("plate$well_row is not a factor. Automatically generating plate$well_row factor levels. May lead to incorrect plate plans.")
        plate <- plate %>%
            dplyr::mutate(well_row = as_factor(.data$well_row))
    }
    
    if (!is.null(colkey)) {
        assertthat::assert_that(assertthat::has_name(colkey, "well_col"))
        # Note: should this if clause be a freestanding function?
        # coerce_column_to_factor(df, col, warn=FALSE)?
        if (!is.factor(colkey$well_col) & coercefactors) {
            message("coercing well_col to a factor with levels from plate$well_col")
            colkey <- dplyr::mutate(
                colkey,
                well_col = factor(.data$well_col,
                                  levels = levels(plate$well_col))
            )
        }
        plate <- dplyr::left_join(plate, colkey, by = "well_col")
    }
    if (!is.null(rowkey)) {
        assertthat::assert_that(assertthat::has_name(rowkey, "well_row"))
        if (!is.factor(rowkey$well_row) & coercefactors) {
            message("coercing well_row to a factor with levels from plate$well_row")
            rowkey <- dplyr::mutate(
                rowkey,
                well_row = factor(.data$well_row,
                                  levels = levels(plate$well_row))
            )
        }
        plate <- dplyr::left_join(plate, rowkey, by = "well_row")
    }
    # check that plate contains sample_id, target_id, prep_type
    if (! "sample_id" %in% names(plate)) {
        message("plate does not contain variable sample_id")
    }
    if (! "target_id" %in% names(plate)) {
        message("plate does not have variable target_id")
    }
    if (! "prep_type" %in% names(plate)) {
        message("plate does not have variable prep_type")
    }
    return(dplyr::arrange(plate, .data$well_row, .data$well_col))
}


#' Display an empty plate plan which can be populated with 
#' ggplot2 geom elements.
#'
#' @param plate tibble with variables well_col, well_row.
#'
#' @return ggplot object; major output is to plot it
#'
#' @examples
#' library(ggplot2)
#' 
#' # display empty plot of empty plate
#' display_plate(create_blank_plate_96well())
#' 
#' # display wells of empty plate filled by column
#' display_plate(create_blank_plate_96well()) + 
#'   geom_tile(aes(fill = well_col), colour = "black")
#' 
#' # display wells of empty 1536-well plate filled by row
#' display_plate(create_blank_plate_1536well()) + 
#'   geom_tile(aes(fill = well_row), colour = "black")
#' 
#' @family plate creation functions
#'
#' @export
#' @importFrom forcats as_factor
#' @importFrom rlang .data
#'
display_plate <- function(plate) {
    assertthat::assert_that(
        assertthat::has_name(plate, 
                             c("well_row","well_col")))
    
    rowlevels <- 
        dplyr::pull(plate, .data$well_row) %>%
        as_factor() %>%
        levels()

    ggplot2::ggplot(data = plate,
                    ggplot2::aes(x = as_factor(.data$well_col),
                        y = as_factor(.data$well_row))) +
        ggplot2::scale_x_discrete(expand = c(0, 0)) +
        ggplot2::scale_y_discrete(expand = c(0, 0),
                                  limits = rev(rowlevels)) +
        ggplot2::coord_equal() +
        ggplot2::theme_void() +
        ggplot2::theme(axis.text = ggplot2::element_text(angle = 0),
                       panel.grid.major = ggplot2::element_blank(),
                       legend.position = "none",
                       plot.margin = grid::unit(rep(0.01, 4), "npc"),
                       panel.border = ggplot2::element_blank())
}

#' Display qPCR plate plan with sample_id, target_id, prep_type per well
#'
#' @param plate tibble with variables well_col, well_row, sample_id, target_id,
#'   prep_type. Output from label_plate_rowcol.
#'
#' @return ggplot object; major output is to plot it
#'
#' @examples 
#' 
#' 
#' # create basic 6-well plate
#' basic_plate <- 
#'     label_plate_rowcol(plate = create_blank_plate(well_row = LETTERS[1:2],
#'                                                   well_col = 1:3),
#'                        rowkey = tibble(well_row = factor(LETTERS[1:2]),
#'                                        target_id = c("T_A","T_B")),
#'                        colkey = tibble(well_col = factor(1:3),
#'                                        sample_id = c("S_1","S_2", "S_3"),
#'                                        prep_type = "+RT"))
#' 
#' # display basic plate
#' display_plate_qpcr(basic_plate)
#' 
#' # create full 384 well plate
#' full_plate <- label_plate_rowcol(create_blank_plate(), 
#'                                   create_rowkey_8_in_16_plain(target_id = c("T_1", "T_2",
#'                                                                             "T_3", "T_4", 
#'                                                                             "T_5", "T_6",
#'                                                                             "T_7", "T_8")), 
#'                                   create_colkey_6diln_2ctrl_in_24() %>% 
#'                                       dplyr::mutate(sample_id = paste0(dilution_nice,
#'                                                                        "_",
#'                                                                        tech_rep)))
#' 
#' # display full plate
#' display_plate_qpcr(full_plate)
#' 
#' @family plate creation functions
#'
#' @export
#' @importFrom rlang .data
#'
display_plate_qpcr <- function(plate) {
    assertthat::assert_that(
        assertthat::has_name(plate, 
                             c("target_id",
                               "sample_id",
                               "prep_type")))
    
    display_plate(plate) +
        ggplot2::geom_tile(ggplot2::aes(fill = .data$target_id), 
                           alpha = 0.3) +
        ggplot2::geom_text(ggplot2::aes(label = 
                                            paste(.data$target_id,
                                                  .data$sample_id,
                                                  .data$prep_type,
                                                  sep = "\n")),
                           size = 2.5, lineheight = 1)
}

#' Display the value of each well across the plate. 
#' 
#' Plots the plate with each well coloured by its value. Example values are Cq, Delta Cq or Delta Delta Cq.
#'
#' For a specific example see the calibration vignette:
#' \code{vignette("calibration_vignette", package = "tidyqpcr")}
#'
#' @param plate tibble with variables well_col, well_row, and the variable to be plotted.
#' 
#' @param value character vector selecting the variable in plate to plot as the well value
#'
#' @return ggplot object; major output is to plot it
#'
#' @examples 
#' library(dplyr)
#' library(ggplot2)
#' 
#' # create 96 well plate with random values
#' plate_randomcq <- create_blank_plate_96well() %>%
#'     mutate(cq = runif(96) * 10,
#'            deltacq = runif(96) * 2)
#' 
#' 
#' # display well Cq value across plate
#' display_plate_value(plate_randomcq)
#' 
#' # display well Delta Cq value across plate with red colour pallette
#' display_plate_value(plate_randomcq, value = "deltacq") +   # uses ggplot syntax
#'     scale_fill_gradient(high = "#FF0000") 
#'           
#' 
#' @family plate creation functions
#'
#' @export
#' @importFrom forcats as_factor
#' @importFrom rlang .data
#'
display_plate_value <- function(plate, value = "cq") {
    # check value exists in given plate
    assertthat::assert_that(value %in% names(plate), 
                            msg = paste0(value, " is not the name of a variable in the given plate"))
    
    # check each well has one value only
    unique_well_value <- plate %>%
        dplyr::group_by(.data$well) %>%
        dplyr::summarise(num_well = dplyr::n()) %>%
        dplyr::mutate(not_equal_one = .data$num_well != 1)
    
    assertthat::assert_that(sum(unique_well_value$not_equal_one) == 0, 
                            msg = paste0("Wells do not have unique ", value, " value."))
    
    rowlevels <- 
        dplyr::pull(plate, .data$well_row) %>%
        as_factor() %>%
        levels()
    
    display_plate(plate = plate) +
        ggplot2::geom_tile(ggplot2::aes(fill = .data[[value]])) +
        ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5),
                       panel.grid.major = ggplot2::element_blank()) +
        ggplot2::labs(title = paste0({{value}}, " values for each well across the plate"))
}
